### ISSUE TITLE
Module.sow() should fail if collection provided as input

### ISSUE THREAD

**Issue by andsteing** [google/flax#1227]
It's very easy to accidentally mis-use `flax.linen.Module.sow()` by calling `apply()` and feeding the intermediate values back in. This results in ever growing:

```python
class Module(nn.Module):
  @nn.compact
  def __call__(self, x):
    self.sow('intermediates', 'x', x)
    return x

x = jnp.array(1.)
mod = Module()
variables = mod.init(jax.random.PRNGKey(0), x)
for i in range(3):
  _, mod_vars = mod.apply(variables, x, mutable='intermediates')  # recompiles!
  variables = variables.copy(mod_vars)

len(variables['intermediates']['x'])  # 4 copies !
```

Note that appending values is desired behaviour because the same module can be called multiple times and we would want to recover all the values, not only the last value. Note also that `variables = variables.copy(mod_vars)` is a relatively common pattern with mutable variable collections.

Ideally `Module.sow()` would detect if any variable is passed in via `.Module.apply()` and then fail instead of silently appending values.

**Comment by marcvanzee** [comment]
Tagging @jheek who implemented this and may also have an opinion on it

**Comment by jheek** [comment]
Feeding back the intermediates is not always a problem:
For example if you use apply inside another module then feeding back the intermediates is probably the right thing to do.
In general you could have a finite number of apply calls that you want to track something on with sow.

Also in the current sow api appending is the default behaviour but users can override this. For example you could also use sow to track a running average or a sum in which case you probably want to feed the sowed values back in.

**Comment by marcvanzee** [comment]
I don't understand this issue actually. @andsteing and @jheek, you both seem to agree that appending is the intended behavior, and the docstring is pretty clear about this as well. To me it also doesn't seem wrong to append the intermediate values if existing ones are found, I can imagine users might want to see the history of all intermediates.

So given that we all agree with this, what is the issue?

**Comment by andsteing** [comment]
oh I think I should have closed this after #1228
